### PR TITLE
feature: Add request/response header support

### PR DESCRIPTION
**Pull request by leoorshansky**
Adds a pre-populated headers struct to the Request object which can be iterated over lazily (using the JS iterator handle).

Adds a headers struct to the Response object which can be modified using with_headers, set_headers or headers_mut.

Headers struct can be constructed from scratch (create new, then set() and append()) or using Iterator<(String, String)>.collect() and Iterator<(&str, &str)>.collect()

Adds an example to demonstrate both of these capabilities.

### CODE REVIEW

**Comment by nilslice** [comment]
@leoorshansky - this is looking great. let's set up some time to review together. I'll ping you.

File: rust-sandbox/src/lib.rs [hunk 1]
```diff
@@ -26,6 +26,15 @@ pub async fn main(mut req: Request) -> Result<Response> {
             );
             Response::ok(Some(msg))
         }
+        (Method::Get, "/headers") => {
+            let msg = req.headers().into_iter()
+                .map(|(name, value)| {
+                    format!("{}: {}\n", name, value)
+                }).collect();
+            let mut headers: worker::Headers = [("Content-Type", "application/json"), ("Set-Cookie", "hello=true")].iter().collect();
+            headers.append("Set-Cookie", "world=true");
```
Comment (nilslice) [inline rust-sandbox/src/lib.rs:35]: Can we add a test to verify the difference between `.append` and `.set`? there might be some good tests we can borrow from other implementations. From MDN, this example would be good to test: 
```js
myHeaders.append('Accept-Encoding', 'deflate');
myHeaders.append('Accept-Encoding', 'gzip');
```

File: worker/src/lib.rs [hunk 2]
```diff
@@ -144,23 +234,41 @@ impl Response {
...
+
+    pub fn set_headers(&mut self, headers: Headers) {
+        self.headers = headers
+    }
+
+    pub fn headers(&self) -> &Headers {
+        &self.headers
+    }
+    
+    pub fn headers_mut(&mut self) -> &mut Headers {
```
Comment (nilslice) [inline worker/src/lib.rs:269]: Can we add a test that proves mutation on the returned `&mut Headers` modifies the underlying Headers owned by the Headers struct?

File: rust-sandbox/src/lib.rs [hunk 3]
```diff
@@ -26,6 +26,15 @@ pub async fn main(mut req: Request) -> Result<Response> {
             );
             Response::ok(Some(msg))
         }
+        (Method::Get, "/headers") => {
+            let msg = req.headers().into_iter()
+                .map(|(name, value)| {
+                    format!("{}: {}\n", name, value)
+                }).collect();
```
Comment (nilslice) [inline rust-sandbox/src/lib.rs:33]: Would be great to show an example of using this through a for-style loop iteration too!

File: worker/src/lib.rs [hunk 4]
```diff
@@ -51,9 +51,92 @@ impl From<(String, u64, String)> for Schedule {
     }
 }
 
+pub struct Headers (EdgeHeaders);
+
+#[allow(clippy::new_without_default)]
+impl Headers {
```
Comment (nilslice) [inline worker/src/lib.rs:57]: Let's handle these unwraps, and maybe change the API to return Result when the underlying JS/Wasm does.

File: .gitignore [hunk 5]
```diff
@@ -1,3 +1,5 @@
 **/target
 Cargo.lock
-.DS_Store
\ No newline at end of file
+.DS_Store
+rust-sandbox/wrangler.toml
```
Comment (nilslice) [inline .gitignore:4]: Let's keep this but `""` set the `account_id` prop

File: worker/src/lib.rs (hunk 4 above)
Comment (nilslice) [inline worker/src/lib.rs:57]: Can we add a `From` impl for the http::HeaderMap type?

File: rust-sandbox/src/lib.rs (hunk 1 above)
Comment (leoorshansky) [reply rust-sandbox/src/lib.rs]: Added tests in worker/tests/headers.rs

File: worker/src/lib.rs (hunk 2 above)
Comment (leoorshansky) [reply worker/src/lib.rs:352]: Added

File: rust-sandbox/src/lib.rs (hunk 3 above)
Comment (leoorshansky) [reply rust-sandbox/src/lib.rs:38]: Added

File: worker/src/lib.rs (hunk 4 above)
Comment (leoorshansky) [reply worker/src/lib.rs:64]: Changed the API

File: .gitignore (hunk 5 above)
Comment (leoorshansky) [reply .gitignore]: done

**Comment by nilslice** [comment]
looks great, @leoorshansky - thanks! what do you think about adding some configurable logging (which would wrap the js console.log)? not for this PR, but just something to start thinking about how we'd implement it.

i will merge this and rebase #3/#5 on it

**Pull request merged**